File: ga/glm4_result_2024_6/result/1968G2_glm/33_1968G2.c
use std::io::{self, BufRead};
use std::cmp::max;

const N: usize = 200000;
const N_: usize = 1 << 18; // N_ = pow2(ceil(log2(N)))

fn max(a: i32, b: i32) -> i32 {
    a.max(b)
}

fn build(aa: &mut Vec<i32>, n: usize) -> Vec<i32> {
    let mut st = vec![0; N_ * 2];
    let mut n_ = 1;
    while n_ < n {
        n_ <<= 1;
    }
    for i in 0..n {
        st[n_ + i] = aa[i];
    }
    for i in (1..n_).rev() {
        st[i] = max(st[i << 1], st[i << 1 | 1]);
    }
    st
}

fn query(st: &Vec<i32>, l: usize, a: i32) -> usize {
    let r = N_ - 1;
    let mut l = l + N_;
    let mut r = r + N_;
    while l <= r {
        if l & 1 == 1 {
            if st[l] >= a {
                while l < N_ {
                    l = if st[l << 1] >= a { l << 1 } else { l << 1 | 1 };
                }
                return l - N_;
            }
            l += 1;
        } else {
            l >>= 1;
            r >>= 1;
        }
    }
    N_
}

fn main() {
    let stdin = io::stdin();
    let mut lines = stdin.lock().lines();
    let t: usize = lines.next().unwrap().unwrap().parse().unwrap();

    for _ in 0..t {
        let mut line = lines.next().unwrap().unwrap();
        let n: usize = line.parse().unwrap();
        let kl: usize = line.parse().unwrap();
        let kr: usize = line.parse().unwrap();
        let mut cc = String::new();
        line = lines.next().unwrap().unwrap();
        cc.push_str(&line);

        let mut zz = vec![0; n];
        let mut l = 0;
        let mut r = 0;
        for i in 1..n {
            if i + zz[i - l] < r {
                zz[i] = zz[i - l];
            } else {
                l = i;
                r = max(r, i);
                while r < n && cc.chars().nth(r) == cc.chars().nth(r - l) {
                    r += 1;
                }
                zz[i] = r - l;
            }
        }

        let st = build(&mut zz, n);
        let mut ans = vec![0; n + 1];
        for p in 1..=n {
            let mut k = 0;
            let mut i = 0;
            while i < n {
                i = query(&st, i + p, p as i32);
                k += 1;
            }
            ans[k] = max(ans[k], p);
        }

        for k in (1..n).rev() {
            ans[k] = max(ans[k], ans[k + 1]);
        }

        for k in kl..=kr {
            println!("{} ", ans[k]);
        }
    }
}

File: ga/glm4_result_2024_6/result/1968G2_cpp_glm/33_1968G2.rs
use std::io::{self, BufRead};
use std::cmp::Ordering;

const N: usize = 2e5 as usize + 3;

fn find(p: &mut Vec<usize>, k: usize) -> usize {
    if k > N || p[k] == k {
        k
    } else {
        p[k] = find(p, p[k]);
        p[k]
    }
}

fn main() {
    let stdin = io::stdin();
    let mut lines = stdin.lock().lines();

    let mut t = 0;
    while let Some(Ok(line)) = lines.next() {
        let mut input = line.split_whitespace();
        t = input.next().unwrap().parse::<usize>().unwrap();

        while t > 0 {
            let n = input.next().unwrap().parse::<usize>().unwrap();
            let x = input.next().unwrap().parse::<usize>().unwrap();
            let y = input.next().unwrap().parse::<usize>().unwrap();
            let a = input.next().unwrap();

            let mut p = vec![0; N];
            let mut t = (0..=n).collect::<Vec<usize>>();
            let mut ans = vec![0; N];
            let mut z = vec![0; N];

            z[0] = n;

            for i in 1..=n {
                let mut j = 0;
                while j + z[j] > i {
                    z[i] = std::cmp::min(z[i - j + 1], j + z[j] - i);
                }
                while a.chars().nth(i + z[i] - 1) == a.chars().nth(z[i]) {
                    z[i] += 1;
                }
                if i + z[i] > j + z[j] {
                    j = i;
                }
            }

            t.sort_by(|&i, &j| z[i].cmp(&z[j]));

            for i in 1..=n {
                let mut k = 1;
                while k <= n && z[t[k]] < i {
                    p[t[k]] = t[k] + 1;
                    k += 1;
                }
                let mut c = 0;
                for j in 1..=n {
                    c += 1;
                    while find(&mut p, j + i) != j {
                        j = find(&mut p, j + i);
                    }
                }
                ans[c] = i;
            }

            for i in (1..=n).rev() {
                ans[i] = std::cmp::max(ans[i], ans[i + 1]);
            }

            for i in x..=y {
                print!("{} ", ans[i]);
            }
            println!("");
            t -= 1;
        }
    }
}